File: Source/Entities/Controllers/PlayerGroundedFlagController.cs
using Microsoft.Xna.Framework;
using Monocle;
using Celeste.Mod.Entities;

namespace Celeste.Mod.Microlith57Misc.Entities;

[CustomEntity("Microlith57Misc/PlayerGroundedFlagController")]
public sealed class PlayerGroundedFlagController : Entity {

    public string Flag;
    public bool InvertFlag;

    public PlayerGroundedFlagController(
        EntityData data, Vector2 offset
    ) : base(data.Position + offset) {
        this.SetDepthAndTags(data);
        Flag = data.Attr("flag", "playerGrounded");
        InvertFlag = data.Bool("invertFlag");
    }

    private bool Grounded => Scene.Tracker.GetEntity<Player>()?.OnGround() ?? false;

    public override void Update() {
        base.Update();
        (Scene as Level)!.Session.SetFlag(Flag, Grounded ^ InvertFlag);
    }

}

File: Source/Entities/PlayerGroundedFlagController/PlayerGroundedFlagController.lua
local self = mu.controller {
  "PlayerGroundedFlagController",
  name = "Flag if Player Grounded Controller",
  desc = "Detects if the player is on the ground."
}

self.flag "playerGrounded"
  :nonempty()
  :desc "Flag to set when the player is on the ground."
self.invertFlag(false)

return self()
